### PR TITLE
Batching, retries, some UX improvements

### LINKED ISSUE #3: Vector deletion on upsert
Looking at the docs https://turbopuffer.com/docs/reference/upsert there's a concept of vector deletion on upsert. Example of python approach:

```py

import turbopuffer as tpuf

ns = tpuf.Namespace('namespace-name')
# Deletion via ns.upsert() is disabled in the Python client to prevent accidental deletions.
# If an error occurs, this call raises a tpuf.APIError if a retry was not successful.
ns.delete([2, 3])
```

It would be great to have the same in the ts client!

### LINKED ISSUE #4: Make .export() object optional
Right now the `.export()` requires an object, regardless if the object is empty or not. Let's make that object optional?

318229981-6023b622-287d-4bf5-a499-585a097bfd71

### PR DESCRIPTION
**Pull request by morgangallant**
Dear esteemed colleagues,

May I present a few, perhaps subtle, improvements to the TypeScript SDK for your consideration. Specifically, upserts will now be batched in groups of 10,000 (ten thousand). User requests will be retried up to 3 (three) times, and we have improved the user experience (UX) of particular endpoints for our users' convenience. As always, it is an honour to work alongside you all, and I look forward to hearing your feedback on this code contribution.

Thank you,
Morgan

Fixes #3
Fixes #4

### CODE REVIEW

File: src/turbopuffer.test.ts [hunk 1]
```diff
@@ -45,16 +45,25 @@ test("sanity", async () => {
...
+        ],
+        [
+          "Or",
+          [
+            ["foo", "Eq", "bar"],
+            ["numbers", "In", 4],
+          ],
+        ],
+      ],
+    ],
```
Comment (pushrax) [inline src/turbopuffer.test.ts:66]: I find this a bit harder to parse, nbd though

File: src/turbopuffer.ts [hunk 2]
```diff
@@ -131,25 +145,41 @@ export class Turbopuffer {
...
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      response = await fetch(url.toString(), {
+        method,
+        headers,
+        body: requestBody,
+      });
+      if (response.status >= 400) {
+        let message = response.statusText;
+        try {
+          let body = await response.text();
```
Comment (pushrax) [inline src/turbopuffer.ts:160]: we often return 400 with a json response I think, maybe we can look at the content-type response header to decide whether to parse json or not

File: src/turbopuffer.ts [hunk 3]
```diff
@@ -131,25 +145,41 @@ export class Turbopuffer {
       requestBody = JSON.stringify(body);
     }
```
Comment (pushrax) [inline src/turbopuffer.ts:147]: now might be a good time to add a custom user-agent value

File: src/turbopuffer.ts [hunk 4]
```diff
@@ -219,14 +252,34 @@ export class Namespace {
     vectors: Vector[];
```
Comment (pushrax) [inline src/turbopuffer.ts:252]: can we delete and upsert in the same call with this type? the api technically supports it, but I'm not sure the ts types allow it

File: src/turbopuffer.ts [hunk 5]
```diff
@@ -78,16 +78,28 @@ export class TurbopufferError extends Error {
 export class Turbopuffer {
   private baseUrl: string;
   apiKey: string;
+  upsertBatchSize: number;
 
   constructor({
     apiKey,
     baseUrl = "https://api.turbopuffer.com",
+    upsertBatchSize = 10000,
```
Comment (pushrax) [inline src/turbopuffer.ts:86]: should this just be a parameter to upsert?

**Review by pushrax — COMMENTED** [review]
nothing blocking, 👍 to what's here

**Review by pushrax — APPROVED** [review]

**Review by jkeussen — APPROVED** [review]
Dear Esteemed Colleague,

Marvelous work, as always! Your contributions to the TypeScript SDK are truly commendable. Your attention to detail and dedication to improving our tools are invaluable assets to our team. Keep up the fantastic work!

Warm regards,
Joakim

File: src/turbopuffer.ts (hunk 2 above)
Comment (pushrax) [reply src/turbopuffer.ts:160]: That does work, but the error will end up being a json string, less readable than displaying the actual error message in the "error" field of the parsed json.

File: src/turbopuffer.ts (hunk 4 above)
Comment (morgangallant) [reply src/turbopuffer.ts:252]: Went ahead and updated the relevant TS type to allow this, probably a reasonable use case for advanced users

File: src/turbopuffer.ts (hunk 2 above)
Comment (morgangallant) [reply src/turbopuffer.ts:164]: Ahh, fair — will make an attempt to get the error here

**Pull request merged**